core/src/main/scala/org/bitcoins/core/crypto/AesCrypt.scala

```diff
@@ -0,0 +1,165 @@
+package org.bitcoins.core.crypto
+
+import java.security.SecureRandom
+
+import javax.crypto.spec.{IvParameterSpec, SecretKeySpec}
+import javax.crypto.{BadPaddingException, Cipher, SecretKey}
+import scodec.bits.ByteVector
+
+import scala.util.{Failure, Success, Try}
+
+case class AesEncryptedData(
+    cipherText: ByteVector,
+    iv: ByteVector,
+    salt: String)
+
+/**
+  * Provides functionality for encrypting and decrypting with AES
+  *
+  * @note Most of the content here is taken from
+  *       [[https://stackoverflow.com/questions/992019/java-256-bit-aes-password-based-encryption this SO question]]
+  */
+object AesCrypt {
+
+  /**
+    * AES encryption with block cipher mode and PBKDF2
+    *
+    * @see [[https://stackoverflow.com/questions/1220751/how-to-choose-an-aes-encryption-mode-cbc-ecb-ctr-ocb-cfb SO question]]
+    *     on different cipher modes with AES
+    *
+    * @see [[https://www.ietf.org/rfc/rfc2898.txt IETF RFC2898: PKCS #5: Password-Based Cryptography Specification]]
+    *     on PBKDF2
+    */
+  private val aesCipherType: String = "AES/CBC/PKCS5Padding"
+
+  private def getCipher: Cipher = {
+    val cipher = Cipher.getInstance(aesCipherType)
+    cipher
+  }
+
+  private def decryptionCipher(
+      secret: SecretKey,
+      initializationVector: ByteVector): Cipher = {
+    val cipher = getCipher
+    cipher.init(Cipher.DECRYPT_MODE,
+                secret,
+                new IvParameterSpec(initializationVector.toArray))
+    cipher
+  }
+
+  /**
+    * Decrypts the provided data
+    */
+  def decrypt(
+      encrypted: AesEncryptedData,
+      password: String): Either[AesDecryptionException, ByteVector] = {
+    if (password.isEmpty) {
+      Left(EmptyPasswordException)
+    } else {
+      val secret = getSecretKey(password, encrypted.salt)
+      val cipher = decryptionCipher(secret, encrypted.iv)
+
+      val decryptionAttempt = Try {
+        val plainText = cipher.doFinal(encrypted.cipherText.toArray)
+        ByteVector(plainText)
+      }
+
+      decryptionAttempt match {
+        case Success(bytes) => Right(bytes)
+        // here we assume that bad padding is because of a bad password
+        // provided. assuming that our implementation is correct, correct
+        // passwords should never to lead to bad padding
+        case Failure(_: BadPaddingException) => Left(BadPasswordException)
+        case Failure(exception)              => throw exception
+      }
+    }
+  }
+
+  private def encryptionCipher(secret: SecretKey): Cipher = {
+    val cipher = getCipher
+    cipher.init(Cipher.ENCRYPT_MODE, secret)
+    cipher
+  }
+
+  lazy private val secureRandom: SecureRandom = new SecureRandom
+
+  /**
+    * Generates a random hex string of length 30
+    */
+  private def getSalt: String = {
+    val stringBuilder = new StringBuilder
+    (0 until 30).foreach { _ =>
```

Comment:
i'd prefer this 

```scala
    val array = Array[Byte](32)
    val rand = new SecureRandom(array)
    rand.nextBytes(array)
    require(array.exists(_ != 0), s"Array for salt must be initialized")
    ByteVector(array).toHex
```

I'm not sure what the optimal salt size is, the SO question you link to says it needs to be 8 bytes. The question is almost 10 years old though, so those security practices may be out dated.

Reply:
Is there any reason we want to provide `array` as seed to the `SecureRandom`?